File: 자프 프로젝트 준식/c++ 파일/path_logger_node.cpp
#include <chrono>
#include <cmath>
#include <fstream>
#include <memory>
#include <string>
#include <vector>
#include <filesystem>

#include "rclcpp/rclcpp.hpp"
#include "nav_msgs/msg/path.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"

#include "tf2_ros/transform_listener.h"
#include "tf2_ros/buffer.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2/utils.h"

using namespace std::chrono_literals;

class PathLoggerNode : public rclcpp::Node
{
public:
    PathLoggerNode()
        : Node("path_logger_node")
    {
        path_pub_ = this->create_publisher<nav_msgs::msg::Path>("/raw_path", 10);

        tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

        timer_ = this->create_wall_timer(50ms, std::bind(&PathLoggerNode::timerCallback, this));

        path_msg_.header.frame_id = "map";

        min_dist_ = 0.05;
        last_x_ = std::numeric_limits<double>::quiet_NaN();
        last_y_ = std::numeric_limits<double>::quiet_NaN();

        csv_path_ = "/home/misys/shared_dir/raw_path.csv";
        std::filesystem::create_directories(std::filesystem::path(csv_path_).parent_path());

        RCLCPP_INFO(this->get_logger(), "PathLoggerNode started");
    }

    ~PathLoggerNode()
    {
        saveCSV();
    }

private:
    void timerCallback()
    {
        geometry_msgs::msg::TransformStamped tf_msg;
        try {
            tf_msg = tf_buffer_->lookupTransform(
                "map", "base_link", tf2::TimePointZero, 100ms);
        }
        catch (tf2::TransformException& ex) {
            RCLCPP_DEBUG(this->get_logger(), "TF error: %s", ex.what());
            return;
        }

        double x = tf_msg.transform.translation.x;
        double y = tf_msg.transform.translation.y;

        tf2::Quaternion q(
            tf_msg.transform.rotation.x,
            tf_msg.transform.rotation.y,
            tf_msg.transform.rotation.z,
            tf_msg.transform.rotation.w
        );
        double yaw = tf2::getYaw(q);

        // 최소 이동 거리 체크
        if (!std::isnan(last_x_)) {
            double dist = std::hypot(x - last_x_, y - last_y_);
            if (dist < min_dist_)
                return;
        }

        last_x_ = x;
        last_y_ = y;

        geometry_msgs::msg::PoseStamped ps;
        ps.header.frame_id = "map";
        ps.header.stamp = this->get_clock()->now();
        ps.pose.position.x = x;
        ps.pose.position.y = y;
        ps.pose.position.z = 0.0;
        ps.pose.orientation = tf_msg.transform.rotation;

        path_msg_.header.stamp = ps.header.stamp;
        path_msg_.poses.push_back(ps);

        points_.push_back({ x, y, yaw });

        path_pub_->publish(path_msg_);
    }

    void saveCSV()
    {
        RCLCPP_INFO(this->get_logger(), "Saving raw_path to %s", csv_path_.c_str());

        std::ofstream file(csv_path_);
        if (!file.is_open()) {
            RCLCPP_ERROR(this->get_logger(), "Failed to write CSV!");
            return;
        }

        file << "x,y,yaw\n";
        for (const auto& p : points_) {
            file << p[0] << "," << p[1] << "," << p[2] << "\n";
        }

        file.close();
    }

private:
    rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr path_pub_;
    rclcpp::TimerBase::SharedPtr timer_;

    std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;

    nav_msgs::msg::Path path_msg_;

    std::vector<std::array<double, 3>> points_;
    double min_dist_;
    double last_x_;
    double last_y_;

    std::string csv_path_;
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PathLoggerNode>();

    try {
        rclcpp::spin(node);
    }
    catch (...) {}

    rclcpp::shutdown();
    return 0;
}

File: 자프 프로젝트 준식/path_logger_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
path_logger_node.py
- LiDAR + SLAM → TF(map→base_link)에서 포즈를 받아서 raw_path를 만든다.
- nav_msgs/Path 로 /raw_path 토픽에 publish
- 동시에 CSV로 /home/misys/shared_dir/raw_path.csv 에 저장
"""

import rclpy
from rclpy.node import Node
from nav_msgs.msg import Path
from geometry_msgs.msg import PoseStamped
from tf2_ros import Buffer, TransformListener
from tf_transformations import euler_from_quaternion
import csv
import os
import math
from rclpy.time import Time

class PathLoggerNode(Node):
    def __init__(self):
        super().__init__("path_logger_node")
        self.path_pub = self.create_publisher(Path, "/raw_path", 10)

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        self.timer = self.create_timer(0.05, self.timer_cb)  # 20 Hz

        self.path_msg = Path()
        self.path_msg.header.frame_id = "map"

        self.points = []  # (x, y, yaw)

        # 최소 이동 거리(노이즈 방지)
        self.min_dist = 0.05
        self.last_x = None
        self.last_y = None

        self.csv_path = "/home/misys/shared_dir/raw_path.csv"
        os.makedirs(os.path.dirname(self.csv_path), exist_ok=True)

    def timer_cb(self):
        # TF: map → base_link
        try:
            now = rclpy.time.Time()
            trans = self.tf_buffer.lookup_transform(
                "map", "base_link", now, rclpy.duration.Duration(seconds=0.1)
            )
        except Exception as e:
            self.get_logger().debug(f"TF error: {e}")
            return

        x = trans.transform.translation.x
        y = trans.transform.translation.y
        q = trans.transform.rotation
        yaw = euler_from_quaternion([q.x, q.y, q.z, q.w])[2]

        if self.last_x is not None:
            dist = math.hypot(x - self.last_x, y - self.last_y)
            if dist < self.min_dist:
                return

        self.last_x = x
        self.last_y = y

        ps = PoseStamped()
        ps.header.frame_id = "map"
        ps.header.stamp = self.get_clock().now().to_msg()
        ps.pose.position.x = x
        ps.pose.position.y = y
        ps.pose.position.z = 0.0
        ps.pose.orientation = q

        self.path_msg.header.stamp = ps.header.stamp
        self.path_msg.poses.append(ps)
        self.points.append((x, y, yaw))

        self.path_pub.publish(self.path_msg)

    def destroy_node(self):
        # 종료 시 CSV 저장
        self.get_logger().info(f"Saving raw_path to {self.csv_path}")
        with open(self.csv_path, "w", newline="") as f:
            writer = csv.writer(f)
            writer.writerow(["x", "y", "yaw"])
            for x, y, yaw in self.points:
                writer.writerow([x, y, yaw])
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = PathLoggerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
